NumLib/Extrapolation/LocalLinearLeastSquaresExtrapolator-impl.h

```diff
@@ -0,0 +1,140 @@
+
+    namespace PH = std::placeholders;
+    using Self = LocalLinearLeastSquaresExtrapolator<
+        GlobalVector, PropertyTag, LocalAssembler>;
+
+    AssemblerLib::SerialExecutor::executeDereferenced(
+        std::bind(&Self::extrapolateElement, this, PH::_1, PH::_2, PH::_3, PH::_4),
+        local_assemblers, property, *counts);
+
+    _nodal_values.componentwiseDivide(*counts); // TODO BLAS?
+}
+
+template<typename GlobalVector, typename PropertyTag, typename LocalAssembler>
+void
+LocalLinearLeastSquaresExtrapolator<GlobalVector, PropertyTag, LocalAssembler>::
+calculateResiduals(LocalAssemblers const& local_assemblers,
+                   PropertyTag const property)
+{
+    _residuals.resize(local_assemblers.size()); // TODO resize is problematic
+
+    AssemblerLib::SerialExecutor::executeDereferenced(
+        [this, property](std::size_t const i, LocalAssembler const& local_assembler)
+        {
+            _residuals.set(i, calculateResiudalElement(i, local_assembler, property));
+        },
+        local_assemblers);
+
+}
+
+template<typename GlobalVector, typename PropertyTag, typename LocalAssembler>
+void
+LocalLinearLeastSquaresExtrapolator<GlobalVector, PropertyTag, LocalAssembler>::
+extrapolateElement(std::size_t const element_index,
+                   LocalAssembler const& loc_asm, PropertyTag const property,
+                   GlobalVector& counts
+        )
+{
+    auto const& gp_vals = loc_asm.getIntegrationPointValues(
+            property, _integration_point_values_cache);
+
+    const unsigned nn = loc_asm.getShapeMatrix(0).rows(); // number of nodes in the element
+    const unsigned ni = gp_vals.size();                   // number of gauss points in the element
+
+    auto& N = _local_matrix_cache; // TODO make that local?
+    N.resize(nn, ni); // TODO: might reallocate very often
+
+    for (unsigned gp=0; gp<ni; ++gp)
+    {
+        auto const& shp_mat = loc_asm.getShapeMatrix(gp);
+        assert(shp_mat.rows() == nn);
+
+        // copy shape matrix to extrapolation matrix columnwise
+        N.block(0, gp, nn, 1) = shp_mat;
+    }
+
+    // TODO make gp_vals an Eigen::VectorXd const& ?
+    Eigen::Map<const Eigen::VectorXd> const gpvs(gp_vals.data(), gp_vals.size());
+
+    // TODO optimization: Store LU decomposition of N*N^T for reuse?
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1FullPivLU.html
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LLT.html
+    //                    cf. http://eigen.tuxfamily.org/dox/classEigen_1_1LDLT.html
+    //                    Extrapolate several values at once?
+
+    // do the least squares computation
+    Eigen::VectorXd tmp = (N*N.transpose()).ldlt().solve(N*gpvs);
+
+    // TODO: for now always zeroth component is used
+    auto const& indices = _local_to_global(element_index, 0).rows;
+
+    _nodal_values.add(indices, tmp); // TODO does that give rise to PETSc problems?
```

Comment:
My comment was more related to negative indices. I think, they will pose some problems.